### PR TITLE
[New] `no-duplicates`: support inline type import with `prefer-inline` option

### PR DESCRIPTION
**Pull request by snewcomer**
close https://github.com/import-js/eslint-plugin-import/issues/2470

ref https://github.com/typescript-eslint/typescript-eslint/pull/5050

Typescript `import { type A } from 'util/cool-linter'` was introduced in 4.5 https://devblogs.microsoft.com/typescript/announcing-typescript-4-5/#type-on-import-names

This PR seeks to add TypeScript support to the existing rule `no-duplicates`.

```
❌ Invalid `["error", "prefer-inline"]`

import { AValue, type AType } from './mama-mia'
import type { BType } from './mama-mia'

✅ Valid with `["error", "prefer-inline"]`

import { AValue, type AType, type BType } from './mama-mia'
```

### CODE REVIEW

File: docs/rules/no-duplicates.md
Comment (snewcomer) [inline docs/rules/no-duplicates.md:71]: > consistent with this [PR](https://github.com/import-js/eslint-plugin-import/pull/2473/files), we could use `prefer-inline`

File: package.json
Comment (snewcomer) [inline package.json:95]: > Will wait for this PR to get tests passing

https://github.com/import-js/eslint-plugin-import/pull/2473/files#diff-f4543b3f6041f0b4eb2b25d9f4e6030faf96ccad7d8fb3a5209c80eea0b59c41

**Comment by jasongerbes** [comment]
Would it be possible to extend this rule to also test exports?

**Comment by snewcomer** [comment]
@jasongerbes Great q! I'm not seeing inline export type throw any warnings so I'm guessing it is possible!  I didn't see any mention in the release notes.  Either it is an option or a different rule.  After this PR is merged, I imagine we could consider it then.  I don't have an answer right now.

```
export { type A };
```

**Comment by snewcomer** [comment]
cc @ljharb @bradzacher This is ready for your review.

File: src/rules/no-duplicates.js
Comment (ljharb) [inline src/rules/no-duplicates.js:117]:
```suggestion
        const isTypeSpecifier = specifier.importNode.importKind === 'type';
        const inlineTypeImport = context.options[0] && context.options[0]['inlineTypeImport'];
        const insertText = `${inlineTypeImport && isTypeSpecifier ? 'type ' : ''}${specifier.text}`;
```

File: tests/src/rules/no-duplicates.js
Comment (ljharb) [inline tests/src/rules/no-duplicates.js:530]: it'd be nice to use `[].concat()` and a ternary (adding the test cases or `[]`) rather than a totally separate `ruleTester.run`, but it's not a blocker for the PR.

**Review by ljharb — COMMENTED** [review]
Looks good!

**Comment by snewcomer** [comment]
Hi @ljharb!  Just seeing if I could get another friendly review when/if you have time!

File: src/rules/no-duplicates.js
Comment (ljharb) [inline src/rules/no-duplicates.js:112]:
```suggestion
        const inlineTypeImport = context.options[0] && context.options[0].inlineTypeImport;
```

File: src/rules/no-duplicates.js
Comment (ljharb) [inline src/rules/no-duplicates.js:299]:
```suggestion
      }
      if (n.specifiers.some((spec) => spec.importKind === 'type')) {
```

**Review by ljharb — COMMENTED** [review]
Looks good! I'm comfortable with this option name, but `preferInline` would work as well.

What about someone who supports inline type imports, but wants to force non-inline ones? Perhaps this should be an option with a value of "separate" versus "inline"?

**Comment by snewcomer** [comment]
@ljharb Thanks for the reviews and comments!  So this PR is part of a larger effort.

1. fixes a Type import to Inline-Type via `prefer-inline` or `prefer-top-level` - https://github.com/import-js/eslint-plugin-import/pull/2473. 
2. Fixes a Value import to Inline Type - https://github.com/typescript-eslint/typescript-eslint/pull/5050.
3. (this PR) adds support for resolving duplicate Type imports to Inline Types to a single import source.

I think 1) is what you are referencing, correct?

**Comment by snewcomer** [comment]
I updated to `prefer-inline` to be consistent with https://github.com/import-js/eslint-plugin-import/pull/2473.  Does that seem ok?

**Comment by ljharb** [comment]
Yep, 1 is what I'm referencing

File: tests/src/rules/no-duplicates.js
Comment (snewcomer) [inline tests/src/rules/no-duplicates.js:543]: I want to point out this test.  no-duplicates will collapse these imports like this even before this PR.  Is there a different way we would want to handle this case?  When combined with this [PR](https://github.com/typescript-eslint/typescript-eslint/pull/5050), it would convert `y` to `type y`.  @ljharb

Comment (ljharb) [reply tests/src/rules/no-duplicates.js:543]: Is this a difference in Flow vs TS?

Comment (snewcomer) [reply tests/src/rules/no-duplicates.js:543]: I believe so, but I don't think I fully grok the question as well 😄 so sorry if not a direct answer.  From my very limited knowledge, Flow doesn't have inline type imports.  So this test is a bit obtuse and really just shows that for someone who wrote their imports this way, they still have a path (with other rules) to correct the imports to proper inline types.

Comment (ljharb) [reply tests/src/rules/no-duplicates.js:543]: I thought Flow used to be the only thing that has inline type imports, and TS only recently added them?

Comment (snewcomer) [reply tests/src/rules/no-duplicates.js:543]: Oh!  I'm not familiar with Flow but did some looking.  I'm only seeing top level type imports but I didn't spend a lot of time looking.  And yep, TS added them in 4.5 so this is the last PR in a [bucket](https://github.com/import-js/eslint-plugin-import/pull/2475#issuecomment-1304888131) of 3 PRs to allow codebases to migrate from top level (separate from value imports) type imports to inline type imports.

Comment (ljharb) [reply tests/src/rules/no-duplicates.js:543]: If there’s any autofix that would conflict based on using TS or Flow, then for those we need to use suggestions instead.

Comment (snewcomer) [reply tests/src/rules/no-duplicates.js:543]: One thing to note is that this will only happen if TS > 4.5.  A Flow project would presumably fail upstream if someone tried `import { type x }`.  So I think this test case is not applicable to Flow

Comment (snewcomer) [reply tests/src/rules/no-duplicates.js:543]: @bradzacher Do you have any additional pieces of info to add here?

Comment (ljharb) [reply tests/src/rules/no-duplicates.js:543]: @snewcomer meaning, the rule detect TS 4.5+ and only does this autofix in that case?

Comment (snewcomer) [reply tests/src/rules/no-duplicates.js:543]: More-so you won't even be able to build the project if you use invalid syntax.  In this case below, this is a project TS < 4.5.  I would assume similar things would happen in Flow if using inline types (assuming they don't have support for them at all)

Screen Shot 2022-11-25 at 7 17 23 AM

Comment (ljharb) [reply tests/src/rules/no-duplicates.js:543]: Right - and an eslint autofix should never cause a project to become unbuildable or change semantics. That's what "safe" means in the context of autofixing.

Comment (snewcomer) [reply tests/src/rules/no-duplicates.js:543]: So to clarify before I investigate an addition here - I need to prevent the case where somebody erroneously sets ‘prefer-inline: true’ but their TS version (or Flow) doesn’t support inlining types?

Comment (snewcomer) [reply tests/src/rules/no-duplicates.js:582]: > erroneously sets ‘prefer-inline: true’

To clarify further, meaning they opt-into something that their project does not support.

Comment (snewcomer) [reply tests/src/rules/no-duplicates.js:582]: I'm sorry to bother.  Always waffling whether to bother busy maintainers like yourself.  Do you happen to have a thought here?

> So to clarify before I investigate a new commit to this PR - I need to prevent the case where somebody erroneously sets ‘prefer-inline: true’ but their TS version (or Flow) doesn’t support inlining types?

Comment (snewcomer) [reply tests/src/rules/no-duplicates.js:582]: Thank you @ljharb!  How does [this](https://github.com/import-js/eslint-plugin-import/pull/2475/commits/f495fd931c8b48d300f5ce9d3c7274f2fffb35b6) look?

Comment (ljharb) [reply tests/src/rules/no-duplicates.js:582]: Looks great!

**Review by ljharb — APPROVED** [review]
Looks great!

**Comment by ljharb** [comment]
I'm not worried about the node 6 travis failure (it's due to npm 6 breaking it, and travis not yet updating nvm to work around it) nor about the uncovered line.

File: src/rules/no-duplicates.js
Comment (snewcomer) [inline src/rules/no-duplicates.js:310]: This seems to be the source of error for a few folks

**Pull request merged**
